src/datacollector/data_collector.cc

```diff
@@ -0,0 +1,908 @@
+            if (ec) {
+                LOG(WARNING) << "create hard link failed, src: " << src << ", dst: " << dst
+                             << ", err: " << ec.message();
+                return false;
+            }
+            if (updated) {
+                *updated = true;
+            }
+        }
+    }
+    return true;
+}
+
+// for one tid-pid, SyncOnce is serial(supported by add next task logic), so we don't need to lock for one tid-pid
+void DataCollectorImpl::SyncOnce(uint32_t tid, uint32_t pid) {
+    // read the progress
+    auto task = GetTaskInfo(tid, pid);
+    if (!task.has_tid()) {
+        LOG(WARNING) << "task not exist(deleted), tid: " << tid << ", pid: " << pid << ", reject task";
+        return;
+    }
+    LOG(INFO) << "sync once task: " << task.ShortDebugString();
+    auto name = EncodeId(tid, pid);
+    auto start_point = task.sync_point();
+    auto mode = task.mode();
+
+    // TODO(hw): fetch binlog in a work thread
+    // we fetch binlog even when we read snapshot, cuz we don't want lost any binlog file, should fetch in time
+    std::string binlog_path_in_db;
+    {
+        std::lock_guard<std::mutex> lock(cache_mutex_);
+        binlog_path_in_db = binlog_path_map_[name];
+    }
+    bool updated = false;
+    if (FetchBinlogUnlocked(name, binlog_path_in_db, &updated) == false) {
+        LOG(WARNING) << "fail to fetch binlog, may lost data, reject task: " << task.ShortDebugString();
+        return;
+    }
+    // to avoid too much used binlog
+    auto replicator = GetReplicator(name);
+    if (replicator == nullptr) {
+        LOG(WARNING) << "fail to get replicator, reject task: " << task.ShortDebugString();
+        return;
+    }
+    // only update log part when fetch new, and no need to delete old log files when there is no new log file
+    if (updated) {
+        replicator->Recover();
+        LOG(INFO) << "log parts: " << LogPartsToString(replicator->GetLogPart());
+        if (start_point.type() == datasync::SyncType::BINLOG &&
+            replicator->GetSnapshotLastOffset() < start_point.offset()) {
+            replicator->SetSnapshotLogPartIndex(start_point.offset());
+            // delete old log file if SetSnapshotLogPartIndex
+            // delete will update log parts, it's ok
+            replicator->DeleteBinlog();
+        }
+    }
+
+    // fullfil the request: io_buf, count, next_point, finished(for mode 0)
+    butil::IOBuf io_buf;
+    uint64_t count = 0;
+    decltype(start_point) next_point;
+    bool meet_binlog_end = false;
+    // we can do switch in it
+    if (!PackData(task, &io_buf, &count, &next_point, &meet_binlog_end)) {
+        LOG(WARNING) << "fail to pack data, reject task: " << task.ShortDebugString();
+        return;
+    }
+
+    LOG(INFO) << "pack cnt:" << count << "-size:" << io_buf.size() << ", start " << start_point.ShortDebugString()
+              << ", next: " << next_point.ShortDebugString();
```

Comment:
DLOG?

Reply:
convert to vlog, DLOG needs re-compile, not good for user.